cli/azd/pkg/infra/azure_resource_manager_test.go

```diff
@@ -245,3 +248,96 @@ func TestGetDeploymentResourceOperationsWithNestedDeployments(t *testing.T) {
 	require.Equal(t, 1, subCalls)
 	require.Equal(t, 2, groupCalls)
 }
+
+type listResourceGroupFn = func(context.Context, string, *azcli.ListResourceGroupOptions) ([]azcli.AzCliResource, error)
+
+func TestFindResourceGroupForEnvironment(t *testing.T) {
+
+	t.Parallel()
+
+	// sequence combines multiple steps into one function that can be passed to ListResourceGroupFn. The first invocation
+	// of the returned function called the first step, the second invocation calls the second, and so on. If the returned
+	// function is called more times then there are steps, it panics.
+	sequence := func(steps ...listResourceGroupFn) listResourceGroupFn {
+		i := 0
+		return func(ctx context.Context, subscriptionId string, listOptions *azcli.ListResourceGroupOptions) ([]azcli.AzCliResource, error) {
+			if i >= len(steps) {
+				panic("no step registered")
+			}
+
+			v, err := steps[i](ctx, subscriptionId, listOptions)
+			i++
+			return v, err
+		}
+	}
+
+	tests := []struct {
+		name              string
+		rgsFromFilter     []string
+		rgsFromNoFilter   []string
+		expectedName      string
+		expectedErrorText string
+	}{
+		{"oneTaggedResourceGroup", []string{"custom-rg-name"}, nil, "custom-rg-name", ""},
+		{"noTagButPrefix", nil, []string{"rg-test-env"}, "rg-test-env", ""},
+		{"noTagButSuffix", nil, []string{"test-env-rg"}, "test-env-rg", ""},
+		{"twoTaggedResourceGroups", []string{"custom-rg-name", "custom-rg-name-2"}, nil, "", "more than one possible resource group was found"},
+		{"noResourceGroups", nil, nil, "", "0 resource groups with prefix or suffix with value"},
+		{"noTagMultipleMatches", nil, []string{"test-env-rg", "rg-test-env"}, "", "more than one possible resource group was found"},
+	}
+
+	for _, tt := range tests {
```

Comment:
From the developer perspective, this is so much better than duplicating all test.
But, from a debugger perspective, this becomes an issue.  If there's a test fail in the middle of the loop, you need to update the `test` array temporarily to run the affected test alone.
During CI, you need to re-run all the loop, there's no way to run just one of the tests.

So, I usually prefer the code duplication, just to help making tests unique.

Reply:
FYI - I've pulled this part out into https://github.com/Azure/azure-dev/pull/844#pullrequestreview-1136537002